use :Man instead of `man` for K

justinmk opened the pull request:
ref https://github.com/neovim/neovim/issues/1468

The `man.vim` ftplugin included with n/vim can be used instead of `man` in the shell. Since neovim doesn't support interactive shell yet (https://github.com/neovim/neovim/issues/1496), this can be used to provide `K` functionality where `man` would have been called instead.

IMO this is better than shelling out to `man` anyway (nice syntax highlighting).

Tasks:
- [X] change the behavior of `K` to call `:Man` instead of shelling out to `man`
- [X] new feature: if `'keywordprg'` begins with ":", it is invoked as a Vim ex command instead of a shell command
- [X] update docs
- [X] convert ftplugin to actual plugin
- [x] interpret `[count]K` as `:[range]Man`
- [x] write tests

justinmk committed:
man.vim: convert ftplugin to actual plugin.

- do not create leader maps
- :norm! instead of :norm
- :keepjumps during layout
- use blackhole reg to avoid polluting unnamed reg
- format buffer name as "man://foo(2)"
- simulate behavior of `man`
    - buffer-local mapping of q to quit
    - open in new tab instead of new window
    - set 'nolist'
    - set tabstop=8

justinmk commented on runtime/ftplugin/man.vim:
```diff
@@ -26,11 +26,7 @@ if &filetype == "man"
   " Add mappings, unless the user didn't want this.
   if !exists("no_plugin_maps") && !exists("no_man_maps")
-    if !hasmapto('<Plug>ManBS')
-      nmap <buffer> <LocalLeader>h <Plug>ManBS
```
removing this misfeature

justinmk commented on runtime/ftplugin/man.vim:
```diff
@@ -41,7 +37,6 @@ endif
 
 if exists(":Man") != 2
   com -nargs=+ Man call s:GetPage(<f-args>)
-  nmap <Leader>K :call <SID>PreGetPage(0)<CR>
```
instead of mapping `<leader>K`, the default `K` behavior should call `:Man` (not done yet).

justinmk commented on runtime/ftplugin/man.vim:
```diff
@@ -143,32 +138,32 @@ func <SID>GetPage(...)
-  setl bufhidden=hide
-  setl nobuflisted
+  setlocal ft=man nomodified nomodifiable readonly bufhidden=hide nobuflisted
+  setlocal tabstop=8 nolist colorcolumn=0
```
some layout tweaks

justinmk commented on runtime/ftplugin/man.vim:
```diff
@@ -143,32 +138,32 @@ func <SID>GetPage(...)
-  setl nobuflisted
+  setlocal ft=man nomodified nomodifiable readonly bufhidden=hide nobuflisted
+  setlocal tabstop=8 nolist colorcolumn=0
+  nnoremap <nowait><buffer> q <c-w>c
```
I'm a little iffy on this. My hunch is that users are used to pressing `q` to exit `man` after `K`, so we should preserve that behavior.

aktau replied:
I most definitely would press q, out of habit.

tarruda commented:
great :+1:

justinmk committed:
'keywordprg': support ex commands

- new feature: if the first character of 'keywordprg' is ":", the
  command is invoked as a Vim ex-command prefixed with [count].
- change default 'keywordprg' to :Man

justinmk commented on src/nvim/normal.c:
```diff
@@ -4388,10 +4386,10 @@ static void nv_ident(cmdarg_T *cap)
     init_history();
-    add_to_history(HIST_SEARCH, buf, true, NUL);
-    normal_search(cap, cmdchar == '*' ? '/' : '?', buf, 0);
+    add_to_history(HIST_SEARCH, (char_u *)buf, true, NUL);
```
I see casts from `(char_u *)` to `(char *)` everywhere, but isn't that a narrowing conversion? Is it safe to cast from `(char *)` to `(char_u *)` (widening conversion)?

aktau replied:
The amount of bytes is the same (and the signedness of char is platform depedent, it's a weird type). It should be absolutely fine. The only thing that might differ is the generated instructions: wether they are signed or unsigned adds/subs/cmps/...

Said another way, it's not as dangerous as casting (uint64_t *) to (uin32_t *).

fwalch commented on runtime/doc/options.txt:
```diff
@@ -4414,19 +4414,21 @@ A jump table for the options with a short description can be found at |Q_op|.
-	When "man" is used, Vim will automatically translate a count for the
-	"K" command to a section number.  Also for "man -s", in which case the
-	"-s" is removed when there is no count.
+	When the first character is ":", the command is invoked as a Vim
```
Add `{Nvim}` somewhere, to track this as difference from Vim?

justinmk replied:
:+1:

lucc commented on runtime/ftplugin/man.vim:
```diff
@@ -143,32 +138,32 @@ func <SID>GetPage(...)
-  setl ma nonu nornu nofen
-  silent exec "norm 1GdG"
+  setlocal modifiable nonumber norelativenumber nofoldenable
+  silent norm! 1GdG
```
Why not simply `%delete`? Why `silent`, can this produce an error?

chrisbra replied:
Am 2015-07-01 09:44, schrieb Lucas Hoffmann:
> In runtime/ftplugin/man.vim [1]:
> 
>> 
>> - setl ma nonu nornu nofen
>> - silent exec "norm 1GdG"
>> + setlocal modifiable nonumber norelativenumber nofoldenable
>> + silent norm! 1GdG
> 
> Why not simply %delete? Why silent, can this produce an error?

It prevents the "No Lines in buffer" message

ZyX-I replied:
Why all deletes here do not use black hole register?

lucc commented on runtime/ftplugin/man.vim:
```diff
@@ -143,32 +138,32 @@ func <SID>GetPage(...)
   while getline('$') =~ '^\s*$'
-    silent norm Gdd
+    silent norm! Gdd
   endwhile
```
Why not use this?
```vim
0,/[^[:space:]]/ global /^\s*$/ delete
silent $,?[^[:space:]]? global /^\s*$/ delete
```

Sean1708 commented:
Better handling of `K` is desperately needed I think, well done. One thing I think would be really good if this was implemented as a function (I've done a similar thing [here](https://github.com/Sean1708/dotfiles/blob/master/neovim/.nvim/autoload/kwp.vim) but it's essentially a very simple hack) making it very simple for people to set their own `K` mapping ([like so](https://github.com/Sean1708/dotfiles/blob/master/neovim/.nvim/after/ftplugin/julia.vim#L20)).

lucc commented:
Dr. Chip also has a [manpageview plugin](http://www.drchip.org/astronaut/vim/index.html#MANPAGEVIEW) which has many advantages over the default man.vim. Maybe you can have a look at it for inspiration or even get him to help.

justinmk commented:
> One thing I think would be really good is if this was implemented as a function

I don't understand. This PR adds support for arbitrary Vim commands to the built-in filetype-specific `'keywordprg'`, so that should cover any possible customization scenario.

justinmk committed:
build: remove USEMAN_S

justinmk committed:
man.vim: man#get_page(): parse page and section.

- Eliminate man#pre_get_page().
- Temporarily remove () from 'iskeyword' to avoid spurious \k match.

justinmk commented on runtime/autoload/man.vim:
```diff
@@ -0,0 +1,131 @@
+" Expects a string like 'access' or 'access(2)'.
+function s:parse_page_and_section(str)
+  let l:page = substitute(a:str, '(*\(\k\+\).*', '\1', '')
+  let l:sect = substitute(a:str, '\(\k\+\)(\([^()]*\)).*', '\2', '')
```
There is an extremely bizarre bug (present in at least Vim 7.3 as well) that causes the result of 

    let l:page = substitute(a:str, '(*\(\k\+\).*', '\1', '')

to return `"foo(1)"` if it is called twice in a row on a `filetype=man` page. Haven't narrow it down yet.

justinmk committed:
man.vim: window-local options

Set window-local options only on a newly-created "man" tab or if we're
already in one.

justinmk committed:
nv_ident: sprintf => snprintf

Also fix some other clint errors.

justinmk committed:
test: K, 'keywordprg'

justinmk commented:
RFC.

justinmk committed:
man.vim: accept [count] for section number.

justinmk commented on runtime/doc/various.txt:
```diff
@@ -542,7 +542,7 @@ which it was defined is reported.
 
 							*K*
-K			Run a program to lookup the keyword under the
+[count]K       		Run a program to lookup the keyword under the
```
tab

justinmk commented on runtime/doc/various.txt:
```diff
@@ -557,12 +557,14 @@ K			Run a program to lookup the keyword under the
+			- When 'keywordprg' is equal to "man", a [count]
+			  before "K" is inserted after the "man" command and
+			  before the keyword.  For example, using "2K" while
+			  the cursor is on "mkdir", results in: >
```
tigten this up

renstrom commented:
I'm not gonna pretend to know everything that goes on in this pull request, but does this supersede/obsolete https://github.com/neovim/neovim/pull/3104?

justinmk commented:
@renstrom No--I would have let you know if it did :)
